File: soccer/gameplay/plays/TheirFreekick.cpp
#include "TheirFreekick.hpp"

#include <iostream>
#include <boost/foreach.hpp>

using namespace std;
using namespace Geometry2d;

REGISTER_PLAY_CATEGORY(Gameplay::Plays::TheirFreekick, "Restarts")

Gameplay::Plays::TheirFreekick::TheirFreekick(GameplayModule *gameplay):
	Play(gameplay),
	_fullback1(gameplay, Behaviors::Fullback::Left),
	_fullback2(gameplay, Behaviors::Fullback::Right),
	_marking1(gameplay),
	_marking2(gameplay)
{
	_fullback1.otherFullbacks.insert(&_fullback2);
	_fullback2.otherFullbacks.insert(&_fullback1);

	// assign general parameters
	float r = 0.3;
	_marking1.ratio(r);
	_marking2.ratio(r);
}

float Gameplay::Plays::TheirFreekick::score ( Gameplay::GameplayModule* gameplay )
{
	const GameState &gs = gameplay->state()->gameState;
	return (gs.setupRestart() && gs.theirFreeKick()) ? 0 : INFINITY;
}

bool Gameplay::Plays::TheirFreekick::run()
{
	set<OurRobot *> available = _gameplay->playRobots();
	
	assignNearest(_fullback1.robot, available, Geometry2d::Point());
	assignNearest(_fullback2.robot, available, Geometry2d::Point());
	//FIXME - How to choose?
	assignNearest(_marking1.robot, available, Geometry2d::Point());
	assignNearest(_marking2.robot, available, Geometry2d::Point());
	
	if (!ball().valid)
	{
		return false;
	}
	Point ballPos = ball().pos;

	//  determine which robots to mark
	map<float, OpponentRobot*> open_opp;
	BOOST_FOREACH(OpponentRobot * r, state()->opp) {
		// want maximum distance that is less than 3 meters from any of our robots
		Point oppPos = r->pos;
		float ballDist = oppPos.distTo(ballPos);
		float max_relevant_robot_range = 3.0; // meters
		if (ballDist < max_relevant_robot_range) {
			float closestSelfDist = 100.0;
			BOOST_FOREACH(OurRobot * s, state()->self) {
				float selfDist = s->pos.distTo(oppPos);
				if (selfDist < closestSelfDist) {
					closestSelfDist = selfDist;
				}
			}
			if (closestSelfDist < 100.0) {
				// record the robot with distance, sort by open-ness
				open_opp[closestSelfDist] = r;
			}
		}
	}

	// assign targets
	if (open_opp.empty()) {
		// if nothing open, just drive at the ball
		//FIXME - They both drive towards the ball?  What?
		if (_marking1.robot)
		{
			_marking1.robot->move(ballPos);
			_marking1.robot->face(ballPos);
		}
		if (_marking2.robot)
		{
			_marking2.robot->move(ballPos);
			_marking2.robot->face(ballPos);
		}
	} else if (open_opp.size() == 1) {
		if (_marking1.robot)
		{
			_marking1.markRobot(open_opp.begin()->second);
			_marking1.run();
		}
		if (_marking2.robot)
		{
			_marking2.robot->move(ballPos);
			_marking2.robot->face(ballPos);
		}
	} else {
		if (_marking1.robot)
		{
			_marking1.markRobot(open_opp.begin()->second);
			_marking1.run();
		}
		if (_marking2.robot)
		{
			_marking2.markRobot((++open_opp.begin())->second);
			_marking2.run();
		}
	}

	// execute default fullback "wall" behavior
	_fullback1.run();
	_fullback2.run();
	
	return true;
}

File: soccer/gameplay2/plays/restarts/their_freekick.py
import play
import behavior
import tactics.positions.fullback
import skills.mark


class TheirFreekick(play.Play):

    def __init__(self):
        super().__init__(continuous=True)

        self.add_transition(behavior.Behavior.State.start,
            behavior.Behavior.State.running,
            lambda: True,
            'immediately')


        fullback1 = tactics.positions.fullback.Fullback(tactics.positions.fullback.Fullback.Side.left)
        self.add_subbehavior(fullback1, 'fullback1', required=False, priority=5)

        fullback2 = tactics.positions.fullback.Fullback(tactics.positions.fullback.Fullback.Side.right)
        self.add_subbehavior(fullback2, 'fullback2', required=False, priority=4)

        self.marks = []
        for i in range(3):
            mark_i = skills.mark.Mark()
            mark_i.ratio = 0.7
            self.add_subbehavior(mark_i, 'mark' + str(i), required=False, priority=3 - i)
            self.marks.append(mark_i)


    def score(self):
        gs = main.game_state()
        return 0 if gs.is_setup_state() and gs.is_their_freekick() else float("inf")


    def execute_running(self):
        # abort if we can't see the ball
        if not main.ball().valid:
            return


        ball_pos = main.ball().pos


        # the closest of their bots to the ball is their kicker
        their_kicker = min(main.their_robots(), key=lambda opp: opp.pos.dist_to(ball_pos))


        # we build an array of (OpponentRobot, float distToClosestOfOurBots) tuples
        # we'll use these in a sec to assign our marking robots
        open_opps_and_dists = []
        for opp in main.their_robots():
            # ignore their kicker
            if opp == their_kicker: continue

            ball_dist = opp.pos.dist_to(ball_pos)

            # see if the opponent is close enough to the ball for us to care
            # if it is, we record the closest distance from one of our robots to it
            if ball_dist < 3.0:
                # which of our robots is closest to this opponent
                closest_self_dist = min([bot.pos.distTo(opp.pos) for bot in main.our_robots()])
                open_opps_and_dists.append( (opp, closest_self_dist) )


        # Decide what each marking robot should do
        # @open_opps contains the robots we want to mark (if any)
        # any robot that isn't assigned a mark_robot will move towards the ball
        # note: subbehaviors are executed before their superbehaviors, so the move() and face() here override the mark behavior's normal control
        for i, mark_i in enumerate(self.marks):
            if mark_i.robot != None:
                if i < len(open_opps_and_dists):
                    # mark the opponent
                    mark_i.mark_robot = open_opps_and_dists[i](0)
                else:
                    # move towards the ball and face it, but don't get within field center's radius
                    pos = mark_i.robot.pos
                    target = pos + (ball_pos - pos).normalized() * (ball_pos.dist_to(pos) - constants.Field.CenterRadius)
                    mark_i.robot.move(target)
                    mark_i.face(ball_pos)


        # tell the marking robots to avoid eachother more than normal
        for i, mark_i in enumerate(self.marks):
            for j, mark_j in enumerate(self.marks):
                if i == j: continue
                if mark_i.robot != None and mark_j.robot != None:
                    mark_i.robot.set_avoid_teammate_radius(mark_j.shell_id(), 0.5)
